### PR TITLE
C3: Use latest version of @cloudflare/workers-types in workers projects

### LINKED ISSUE #4514: C3: Improve `@cloudflare/workers-types` types used for Workers
### Describe the solution

In C3 when creating a new worker we get the latest compatibility date and apply it to the worker's toml file ([source](https://github.com/cloudflare/workers-sdk/blob/04a2d0ed6fca1c366cd891b54026c34e1c1a5701/packages/create-cloudflare/src/workers.ts#L187)) the `@cloudflare/workers-types` types used do not however match such compatibility date (see [comment](https://github.com/cloudflare/workers-sdk/issues/3742#issuecomment-1827092751)) causing a confusing and frustrating experience for developers.

So we should also:
 - make sure that the `@cloudflare/workers-types` is updated to its latest version ~~(instead of having the [version hardcoded](https://github.com/search?q=repo%3Acloudflare%2Fworkers-sdk+%2F%22%40cloudflare%5C%2Fworkers-types%22%3A+%22.*%3F%22%2F+path%3A%2F%5Epackages%5C%2Fcreate-cloudflare%5C%2Ftemplates%5C%2F%2F&type=code))~~ (probably not an issue since the versions do have `^`)
 - update the [tsconfig.json workers-types entries](https://github.com/search?q=repo%3Acloudflare%2Fworkers-sdk+%2F%22types%22%3A+%5C%5B%22%40cloudflare%5C%2Fworkers-types%22%5D%2F+path%3A%2F%5Epackages%5C%2Fcreate-cloudflare%5C%2Ftemplates%5C%2F%2F&type=code) to use the [most recent non-experimental entrypoint](https://github.com/cloudflare/workerd/tree/main/npm/workers-types#compatibility-dates)
 
 This should provide a valid setup for developers and allow them to proceed with their work without type issues.
 
 > [!NOTE]
 > it is still confusing that there is a compatibility date and a (most likely) different workers-types entrpoint date, that is however an issue that needs to be solved separately and not something that we can address in C3, hopefully this will be addressed in the future making the use of types more straightforward for developers

### PR DESCRIPTION
**Pull request by jculvey**
Fixes #4514.

**What this PR solves / how to test:**

Workers projects created via c3 get a stale version of `@cloudflare/workers-types` from their template files and use the default entrypoint in their `tsconfig.json`. This makes for a rough experience for typescript users.

This change fixes that by installing both `wrangler` and `@cloudflare/workers-types` on demand, and by looking up the latest version of the types entrypoint and adjusting the tsconfig.

**Author has addressed the following:**

- Tests
  - [ x ] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ x ] Included
  - [ ] Not necessary because:
- Associated docs
  - [ ] Issue(s)/PR(s):
  - [ ] Not necessary because:

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/create-cloudflare/src/workers.ts [hunk 1]
```diff
@@ -164,33 +162,99 @@ async function copyExistingWorkerFiles(ctx: Context) {
...
 		.replace(/^name\s*=\s*"<TBD>"/m, `name = "${ctx.project.name}"`)
 		.replace(
 			/^compatibility_date\s*=\s*"<TBD>"/m,
 			`compatibility_date = "${await getWorkerdCompatibilityDate()}"`
 		);
+	writeFile(wranglerTomlPath, wranglerToml);
+}
+
+async function installWorkerDependencies(ctx: Context) {
+	const dependencies = ["wrangler"];
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/src/workers.ts:186]:
```suggestion
	const dependencies = ["wrangler@3"];
```

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me 🙂 

The tests look really nice! 🤩👍 

As I mentioned in chat, I am not really convinced that installing `wrangler` and `@cloudflare/workers-types` is the right approach as having them in the `package.json`s with `^` should be enough, but if you strongly believe that it is I am fine with it 🙂

**Comment by jculvey** [comment]
> As I mentioned in chat, I am not really convinced that installing `wrangler` and `@cloudflare/workers-types` is the right approach as having them in the `package.json`s with `^` should be enough, but if you strongly believe that it is I am fine with it 🙂

I tell you what, let's compromise. I'll leave `wrangler` in the templates and keep `@cloudflare/workers-types` where it is. The rationale is that `workers-types` is highly version sensitive at the moment since it's versioned by compat date, and I worry that users of pnpm between 8.0.0 and 8.7.0 will have a bad experience when they get old types (from April) due to the `resolution-mode` issue (see more [here](https://github.com/pnpm/pnpm/releases/tag/v8.7.0)).

**Comment by dario-piotrowicz** [comment]
@jculvey no don't worry I am totally fine with installing both `wrangler` and `@cloudflare/workers-types` I just kind of left the comment to be fully transparent 🙂, but please feel free to do as you prefer I don't think that installing things on the fly is likely to cause issues (it just feels unnecessary to me)

**Pull request merged**